lib/mapping/cache.js

```diff
@@ -0,0 +1,188 @@
+'use strict';
+
+const qModule = require('./q');
+const QueryOperator = qModule.QueryOperator;
+const QueryAssignment = qModule.QueryAssignment;
+
+/**
+ * Provides utility methods for obtaining a caching keys based on the specifics of the Mapper methods.
+ * @ignore
+ */
+class Cache {
+  /**
+   * Gets an iterator of keys to uniquely identify a document shape for a select query.
+   * @param {Array<String>} docKeys
+   * @param {Object} doc
+   * @param {{fields, limit, orderBy}} docInfo
+   * @returns {Iterator}
+   */
+  static *getSelectKey(docKeys, doc, docInfo) {
+    yield* Cache._yieldKeyAndOperators(docKeys, doc);
+
+    if (docInfo) {
```

Comment:
This `if (docInfo)` block seems to be duplicated in `getSelectAllKey` as well, could it be extracted into another method?  Similar story with `getInsertKey` and `getUpdateKey` as well.

Reply:
👍 on `getSelectAllKey` and `getSelectKey` duplication, I'll extract it to another method.

Regarding the other methods, they have very subtle and very important differences between each :)